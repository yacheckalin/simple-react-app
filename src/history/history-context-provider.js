import React, { useContext, useState, useEffect } from "react";

import { data } from "./data";
import { INITIAL_PAGINATION_PAGE, DEFAULT_PAGINATION_LIMIT } from "./constants";
import { customSort } from "./helpers";

export const HistoryContext = React.createContext();
export const useHistoryContext = () => useContext(HistoryContext);

const HistoryContextProvider = ({ children }) => {
  const [paginationList, setPaginationList] = useState([]);
  const [currentPage, setCurrentPage] = useState(INITIAL_PAGINATION_PAGE);
  const [totalPages, setTotalPages] = useState(0);

  useEffect(() => {
    //TODO: change this part for axios request to API
    const successResponse = { result: "ok", deals: data };
    const fetchData = () => Promise.resolve(successResponse);
    let paginated = [];

    const paginate = (data) => {
      return data.splice(
        currentPage * DEFAULT_PAGINATION_LIMIT,
        DEFAULT_PAGINATION_LIMIT
      );
    };

    fetchData().then((response) => {
      if (response.result === "ok") {
        const sorted = customSort(response.deals, currentPage);
        paginated = paginate(sorted);
        setTotalPages(
          Math.floor(response.deals.length / DEFAULT_PAGINATION_LIMIT)
        );

        setPaginationList(paginated);
      }
    });
    return () => setPaginationList(paginated);
  }, [currentPage]);

  const backward = () => {
    setCurrentPage(currentPage - 1);
  };

  const forward = () => {
    setCurrentPage(currentPage + 1);
  };

  return (
    <HistoryContext.Provider
      value={{
        data: paginationList,
        backward,
        forward,
        currentPage,
        totalPages,
      }}
    >
      {children}
    </HistoryContext.Provider>
  );
};

export default HistoryContextProvider;
