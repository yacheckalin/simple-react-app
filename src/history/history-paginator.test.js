import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";

import Pagination from "./history-paginator";
import * as HistoryContext from "./history-context-provider";

describe("<Pagination />", () => {
  test("renders and displays properly", () => {
    const contextValues = {
      currentPage: 0,
      totalPages: 10,
      forward: jest.fn(),
      backward: jest.fn(),
    };
    jest
      .spyOn(HistoryContext, "useHistoryContext")
      .mockImplementation(() => contextValues);

    const wrapper = shallow(<Pagination />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
