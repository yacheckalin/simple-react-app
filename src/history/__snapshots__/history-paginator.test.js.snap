// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> renders and displays properly 1`] = `
<styled.div
  className="pagination-container"
>
  <div
    className="arrow-backward"
    onClick={[MockFunction]}
  />
  <div
    className="numbers"
  >
    1
     / 
    10
  </div>
  <div
    className="arrow-forward"
    onClick={[MockFunction]}
  >
    <IoIosArrowRoundForward
      color="#1a237e"
      size="2em"
    />
  </div>
</styled.div>
`;
